File: solutions/Happy-Number/Happy-Number.js
var cal = (n) => {
  let ans = 0;
  while (n !== 0) {
    const digit = n % 10;
    ans += digit * digit;
    n = Math.floor(n / 10);
  }
  return ans;
};

/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function (n) {
  let slow = n;
  let fast = cal(cal(n));
  while (slow !== fast) {
    slow = cal(slow);
    fast = cal(cal(fast));
  }
  return slow === 1;
};

File: solutions/Happy-Number/Happy-Number.cpp
class Solution {
public:
    bool isHappy(int n) {
        int slow = n;
        int fast = cal(cal(slow));
        while(slow != fast) {
            slow = cal(slow);
            fast = cal(cal(fast));
        }
        return slow == 1;
    }
private:
    int cal(int n) {
        int ans = 0;
        while(n != 0) {
            int digit = n % 10;
            ans += digit * digit;
            n /= 10;
        }
        return ans;
    }
};